File: Cpp/Legacy89DiskKit.Cpp/src/application/event_driven_emulator_fdc_host_adapter.cpp
#include "legacy89diskkit/cpp/application/event_driven_emulator_fdc_host_adapter.hpp"
#include <stdexcept>
#include <algorithm>

namespace legacy89diskkit::cpp::application
{
namespace
{
const EmulatorHostCapabilities HostCapabilities = {
    1,     // protocol_version
    true,  // supports_path_open
    true,  // supports_buffer_open
    true,  // supports_notification_exchange
    true,  // supports_plain_stdio
    true   // supports_observable_stdio
};
}

EventDrivenEmulatorFdcHostAdapter::EventDrivenEmulatorFdcHostAdapter(
    std::shared_ptr<DriveMountService> drive_mount_service,
    std::shared_ptr<MountedMediumBindingService> binding_service)
    : drive_mount_service_(std::move(drive_mount_service)),
      binding_service_(std::move(binding_service))
{
}

void EventDrivenEmulatorFdcHostAdapter::OpenDisk(int drive_number, NativeFileSystemSession& session)
{
    auto binding_result = binding_service_->MountSession(*drive_mount_service_, drive_number, session);
    if (!binding_result.ok())
    {
        throw std::runtime_error(binding_result.status().message);
    }

    auto binding = binding_result.value();
    auto controller_result = binding_service_->CreateController(binding, drive_number);
    if (!controller_result.ok())
    {
        throw std::runtime_error(controller_result.status().message);
    }

    auto controller = controller_result.value();
    auto timed_controller = std::dynamic_pointer_cast<TimedFdcController>(controller);
    if (!timed_controller)
    {
        throw std::runtime_error("The mounted medium controller does not support timing advancement.");
    }

    bindings_[drive_number] = std::move(binding);
    controllers_[drive_number] = ControllerBinding{std::move(controller), std::move(timed_controller)};
    SyncSignals();
}

bool EventDrivenEmulatorFdcHostAdapter::CloseDisk(int drive_number)
{
    bindings_.erase(drive_number);
    controllers_.erase(drive_number);
    bool unmounted = drive_mount_service_->Unmount(drive_number);
    SyncSignals();
    return unmounted;
}

bool EventDrivenEmulatorFdcHostAdapter::IsDiskInserted(int drive_number) const
{
    return drive_mount_service_->IsMounted(drive_number);
}

bool EventDrivenEmulatorFdcHostAdapter::IsDriveReady(int drive_number) const
{
    return drive_mount_service_->GetState(drive_number).is_ready;
}

void EventDrivenEmulatorFdcHostAdapter::SetMotorOn(int drive_number, bool on)
{
    drive_mount_service_->SetMotorOn(drive_number, on);
    SyncSignals();
}

void EventDrivenEmulatorFdcHostAdapter::SelectDrive(int drive_number)
{
    selected_drive_ = drive_number;
    SyncSignals();
}

void EventDrivenEmulatorFdcHostAdapter::SelectSide(int side)
{
    auto it = bindings_.find(selected_drive_);
    if (it == bindings_.end() || !it->second.controller_facing_medium)
    {
        throw std::runtime_error("No controller-facing medium is mounted for the selected drive.");
    }

    it->second.controller_facing_medium->SelectSide(side);
    SyncSignals();
}

void EventDrivenEmulatorFdcHostAdapter::Reset()
{
    GetCurrentController().controller->Reset();
    SyncSignals();
}

void EventDrivenEmulatorFdcHostAdapter::WriteIo8(std::uint32_t address, std::uint8_t value)
{
    GetCurrentController().controller->WriteRegister(MapRegister(address), value);
    SyncSignals();
}

std::uint8_t EventDrivenEmulatorFdcHostAdapter::ReadIo8(std::uint32_t address)
{
    std::uint8_t value = GetCurrentController().controller->ReadRegister(MapRegister(address));
    SyncSignals();
    return value;
}

void EventDrivenEmulatorFdcHostAdapter::Advance(std::chrono::nanoseconds delta)
{
    GetCurrentController().timed_controller->Advance(delta);
    SyncSignals();
}

EmulatorHostResponse EventDrivenEmulatorFdcHostAdapter::Handle(const EmulatorHostRequest& request)
{
    std::optional<std::uint8_t> register_value = std::nullopt;
    std::optional<std::string> error_message = std::nullopt;

    try
    {
        switch (request.kind)
        {
        case EmulatorHostRequestKind::QueryCapabilities:
            break;
        case EmulatorHostRequestKind::OpenDiskPath:
            error_message = "OpenDiskPath is not yet implemented in the C++ Application layer.";
            break;
        case EmulatorHostRequestKind::OpenDiskImage:
            error_message = "OpenDiskImage is not yet implemented in the C++ Application layer.";
            break;
        case EmulatorHostRequestKind::CloseDisk:
            CloseDisk(request.drive_number.value_or(0));
            break;
        case EmulatorHostRequestKind::SelectDrive:
            SelectDrive(request.drive_number.value_or(0));
            break;
        case EmulatorHostRequestKind::SelectSide:
            SelectSide(request.side.value_or(0));
            break;
        case EmulatorHostRequestKind::Reset:
            Reset();
            break;
        case EmulatorHostRequestKind::WriteRegister:
            WriteIo8(request.register_address.value_or(0), request.register_value.value_or(0));
            break;
        case EmulatorHostRequestKind::ReadRegister:
            register_value = ReadIo8(request.register_address.value_or(0));
            break;
        case EmulatorHostRequestKind::Advance:
            Advance(std::chrono::microseconds(request.advance_microseconds.value_or(0)));
            break;
        case EmulatorHostRequestKind::QueryState:
            break;
        }
    }
    catch (const std::exception& e)
    {
        error_message = e.what();
    }

    auto visible_state = TryGetVisibleState();
    auto pending_advance = TryGetPendingAdvanceHint();

    EmulatorHostResponse response;
    response.register_value = register_value;
    response.visible_state = visible_state;
    response.irq = visible_state ? visible_state->irq : false;
    response.drq = visible_state ? visible_state->drq : false;
    response.pending_advance_microseconds = pending_advance ? std::optional<std::int64_t>(std::chrono::duration_cast<std::chrono::microseconds>(*pending_advance).count()) : std::nullopt;
    response.error_message = error_message;
    
    if (request.kind == EmulatorHostRequestKind::QueryCapabilities)
    {
        response.capabilities = HostCapabilities;
    }

    return response;
}

void EventDrivenEmulatorFdcHostAdapter::SyncSignals()
{
    auto visible = TryGetVisibleState();
    bool irq = visible ? visible->irq : false;
    bool drq = visible ? visible->drq : false;
    auto hint = TryGetPendingAdvanceHint();

    if (irq != last_irq_)
    {
        last_irq_ = irq;
        if (on_irq_changed) on_irq_changed(irq);
    }

    if (drq != last_drq_)
    {
        last_drq_ = drq;
        if (on_drq_changed) on_drq_changed(drq);
    }

    if (hint.has_value())
    {
        if (!last_advance_hint_.has_value() || *hint != *last_advance_hint_)
        {
            last_advance_hint_ = hint;
            if (on_advance_requested) on_advance_requested(*hint);
        }
    }
    else
    {
        last_advance_hint_ = std::nullopt;
    }
}

std::optional<FdcVisibleState> EventDrivenEmulatorFdcHostAdapter::TryGetVisibleState() const
{
    auto it = controllers_.find(selected_drive_);
    if (it != controllers_.end())
    {
        return it->second.controller->GetVisibleState();
    }
    return std::nullopt;
}

std::optional<std::chrono::nanoseconds> EventDrivenEmulatorFdcHostAdapter::TryGetPendingAdvanceHint() const
{
    auto it = controllers_.find(selected_drive_);
    if (it != controllers_.end())
    {
        return it->second.timed_controller->GetPendingAdvanceHint();
    }
    return std::nullopt;
}

const EventDrivenEmulatorFdcHostAdapter::ControllerBinding& EventDrivenEmulatorFdcHostAdapter::GetCurrentController() const
{
    auto it = controllers_.find(selected_drive_);
    if (it == controllers_.end())
    {
        throw std::runtime_error("No disk is mounted for the selected drive.");
    }
    return it->second;
}

FdcRegister EventDrivenEmulatorFdcHostAdapter::MapRegister(std::uint32_t address)
{
    switch (address)
    {
    case 0: return FdcRegister::CommandStatus;
    case 1: return FdcRegister::Track;
    case 2: return FdcRegister::Sector;
    case 3: return FdcRegister::Data;
    default: throw std::out_of_range("Unsupported FDC register address.");
    }
}
} // namespace legacy89diskkit::cpp::application

File: CSharp/Legacy89DiskKit/Fdc/Application/Hosts/EventDrivenEmulatorFdcHostAdapter.cs
using Legacy89DiskKit.Drive.Application;
using Legacy89DiskKit.DiskImage.Domain.Interface.Container;
using Legacy89DiskKit.DiskImage.Domain.Interface.Factory;
using Legacy89DiskKit.Fdc.Domain.Interface;
using Legacy89DiskKit.Fdc.Domain.Model;
using Legacy89DiskKit.Fdc.Application.Hosts.Protocol;

namespace Legacy89DiskKit.Fdc.Application.Hosts;

public class EventDrivenEmulatorFdcHostAdapter
{
    private static readonly EmulatorHostCapabilities HostCapabilities = new(
        ProtocolVersion: 1,
        SupportsPathOpen: true,
        SupportsBufferOpen: true,
        SupportsNotificationExchange: true,
        SupportsPlainStdio: true,
        SupportsObservableStdio: true);

    private readonly DriveMountService _driveMountService;
    private readonly MountedMediumBindingService _bindingService;
    private readonly IDiskContainerFactory _containerFactory;
    private readonly Dictionary<int, MountedMediumBinding> _bindings = new();
    private readonly Dictionary<int, ControllerBinding> _controllers = new();
    private readonly Dictionary<int, IDiskContainer> _ownedContainers = new();
    private int _selectedDrive;
    private bool _lastIrq;
    private bool _lastDrq;
    private TimeSpan? _lastAdvanceHint;

    public event Action<bool>? IrqChanged;

    public event Action<bool>? DrqChanged;

    public event Action<TimeSpan>? AdvanceRequested;

    public EventDrivenEmulatorFdcHostAdapter(
        DriveMountService driveMountService,
        MountedMediumBindingService bindingService,
        IDiskContainerFactory containerFactory)
    {
        _driveMountService = driveMountService ?? throw new ArgumentNullException(nameof(driveMountService));
        _bindingService = bindingService ?? throw new ArgumentNullException(nameof(bindingService));
        _containerFactory = containerFactory ?? throw new ArgumentNullException(nameof(containerFactory));
    }

    public void OpenDisk(int driveNumber, IDiskContainer container)
    {
        ArgumentNullException.ThrowIfNull(container);

        var binding = _bindingService.MountContainer(_driveMountService, driveNumber, container);
        var controller = _bindingService.CreateController(binding, driveNumber);
        if (controller is not ITimedFdcController timedController)
        {
            throw new NotSupportedException("The mounted medium controller does not support timing advancement.");
        }

        ReleaseOwnedContainer(driveNumber);
        _bindings[driveNumber] = binding;
        _controllers[driveNumber] = new ControllerBinding(controller, timedController);
        SyncSignals();
    }

    public void OpenDiskPath(int driveNumber, string imagePath, bool readOnly = true)
    {
        if (string.IsNullOrWhiteSpace(imagePath))
        {
            throw new ArgumentException("ImagePath is required.", nameof(imagePath));
        }

        var container = _containerFactory.Open(imagePath, readOnly);
        try
        {
            OpenDisk(driveNumber, container);
            _ownedContainers[driveNumber] = container;
        }
        catch
        {
            container.Dispose();
            throw;
        }
    }

    public void OpenDiskImage(int driveNumber, byte[] imageData, string imageFormat, bool readOnly = true)
    {
        ArgumentNullException.ThrowIfNull(imageData);
        if (string.IsNullOrWhiteSpace(imageFormat))
        {
            throw new ArgumentException("ImageFormat is required.", nameof(imageFormat));
        }

        var container = _containerFactory.Open(imageData, imageFormat, readOnly);
        try
        {
            OpenDisk(driveNumber, container);
            _ownedContainers[driveNumber] = container;
        }
        catch
        {
            container.Dispose();
            throw;
        }
    }

    public bool CloseDisk(int driveNumber)
    {
        _bindings.Remove(driveNumber);
        _controllers.Remove(driveNumber);
        var result = _driveMountService.Unmount(driveNumber);
        ReleaseOwnedContainer(driveNumber);
        SyncSignals();
        return result;
    }

    public bool IsDiskInserted(int driveNumber)
    {
        return _driveMountService.IsMounted(driveNumber);
    }

    public bool IsDriveReady(int driveNumber)
    {
        return _driveMountService.GetState(driveNumber).IsReady;
    }

    public void SelectDrive(int driveNumber)
    {
        EnsureDriveMounted(driveNumber);
        _selectedDrive = driveNumber;
        SyncSignals();
    }

    public void SelectSide(int side)
    {
        if (!_bindings.TryGetValue(_selectedDrive, out var binding) || binding.ControllerFacingMedium is null)
        {
            throw new InvalidOperationException("No controller-facing medium is mounted for the selected drive.");
        }

        binding.ControllerFacingMedium.SelectSide(side);
        SyncSignals();
    }

    public void Reset()
    {
        CurrentController.Controller.Reset();
        SyncSignals();
    }

    public void WriteIo8(uint address, byte value)
    {
        CurrentController.Controller.WriteRegister(MapRegister(address), value);
        SyncSignals();
    }

    public byte ReadIo8(uint address)
    {
        var value = CurrentController.Controller.ReadRegister(MapRegister(address));
        SyncSignals();
        return value;
    }

    public void Advance(TimeSpan delta)
    {
        CurrentController.TimedController.Advance(delta);
        SyncSignals();
    }

    public FdcVisibleState GetVisibleState()
    {
        return CurrentController.Controller.GetVisibleState();
    }

    public bool IsIrqAsserted()
    {
        return GetVisibleState().Irq;
    }

    public bool IsDrqAsserted()
    {
        return GetVisibleState().Drq;
    }

    public EmulatorHostResponse Handle(EmulatorHostRequest request)
    {
        ArgumentNullException.ThrowIfNull(request);

        byte? registerValue = null;

        switch (request.Kind)
        {
            case EmulatorHostRequestKind.QueryCapabilities:
                break;
            case EmulatorHostRequestKind.OpenDiskPath:
                OpenDiskPath(
                    request.DriveNumber ?? throw new ArgumentException("DriveNumber is required.", nameof(request)),
                    request.ImagePath ?? throw new ArgumentException("ImagePath is required.", nameof(request)),
                    request.ReadOnly ?? true);
                break;
            case EmulatorHostRequestKind.OpenDiskImage:
                OpenDiskImage(
                    request.DriveNumber ?? throw new ArgumentException("DriveNumber is required.", nameof(request)),
                    Convert.FromBase64String(request.ImageDataBase64 ?? throw new ArgumentException("ImageDataBase64 is required.", nameof(request))),
                    request.ImageFormat ?? throw new ArgumentException("ImageFormat is required.", nameof(request)),
                    request.ReadOnly ?? true);
                break;
            case EmulatorHostRequestKind.CloseDisk:
                CloseDisk(request.DriveNumber ?? throw new ArgumentException("DriveNumber is required.", nameof(request)));
                break;
            case EmulatorHostRequestKind.SelectDrive:
                SelectDrive(request.DriveNumber ?? throw new ArgumentException("DriveNumber is required.", nameof(request)));
                break;
            case EmulatorHostRequestKind.SelectSide:
                SelectSide(request.Side ?? throw new ArgumentException("Side is required.", nameof(request)));
                break;
            case EmulatorHostRequestKind.Reset:
                Reset();
                break;
            case EmulatorHostRequestKind.WriteRegister:
                WriteIo8(
                    request.RegisterAddress ?? throw new ArgumentException("RegisterAddress is required.", nameof(request)),
                    request.RegisterValue ?? throw new ArgumentException("RegisterValue is required.", nameof(request)));
                break;
            case EmulatorHostRequestKind.ReadRegister:
                registerValue = ReadIo8(request.RegisterAddress ?? throw new ArgumentException("RegisterAddress is required.", nameof(request)));
                break;
            case EmulatorHostRequestKind.Advance:
                Advance(TimeSpan.FromMilliseconds((request.AdvanceMicroseconds ?? throw new ArgumentException("AdvanceMicroseconds is required.", nameof(request))) / 1000.0));
                break;
            case EmulatorHostRequestKind.QueryState:
                break;
            default:
                throw new ArgumentOutOfRangeException(nameof(request), request.Kind, "Unsupported host request kind.");
        }

        var visibleState = TryGetVisibleState();
        var pendingAdvance = TryGetPendingAdvanceHint();

        return new EmulatorHostResponse(
            registerValue,
            visibleState,
            visibleState?.Irq ?? false,
            visibleState?.Drq ?? false,
            pendingAdvance is null ? null : (long)pendingAdvance.Value.TotalMilliseconds * 1000,
            request.Kind == EmulatorHostRequestKind.QueryCapabilities ? HostCapabilities : null);
    }

    private void SyncSignals()
    {
        var visible = TryGetVisibleState();
        var irq = visible?.Irq ?? false;
        var drq = visible?.Drq ?? false;
        var hint = TryGetPendingAdvanceHint();

        if (irq != _lastIrq)
        {
            _lastIrq = irq;
            IrqChanged?.Invoke(irq);
        }

        if (drq != _lastDrq)
        {
            _lastDrq = drq;
            DrqChanged?.Invoke(drq);
        }

        if (hint is null)
        {
            _lastAdvanceHint = null;
            return;
        }

        if (_lastAdvanceHint != hint)
        {
            _lastAdvanceHint = hint;
            AdvanceRequested?.Invoke(hint.Value);
        }
    }

    private FdcVisibleState? TryGetVisibleState()
    {
        if (_controllers.TryGetValue(_selectedDrive, out var controller))
        {
            return controller.Controller.GetVisibleState();
        }

        return null;
    }

    private TimeSpan? TryGetPendingAdvanceHint()
    {
        if (_controllers.TryGetValue(_selectedDrive, out var controller))
        {
            return controller.TimedController.GetPendingAdvanceHint();
        }

        return null;
    }

    private void ReleaseOwnedContainer(int driveNumber)
    {
        if (_ownedContainers.Remove(driveNumber, out var ownedContainer))
        {
            ownedContainer.Dispose();
        }
    }

    private ControllerBinding CurrentController
    {
        get
        {
            if (_controllers.TryGetValue(_selectedDrive, out var controller))
            {
                return controller;
            }

            throw new InvalidOperationException("No disk is mounted for the selected drive.");
        }
    }

    private void EnsureDriveMounted(int driveNumber)
    {
        if (!_controllers.ContainsKey(driveNumber))
        {
            throw new InvalidOperationException("No disk is mounted for the requested drive.");
        }
    }

    private static FdcRegister MapRegister(uint address)
    {
        return address switch
        {
            0 => FdcRegister.CommandStatus,
            1 => FdcRegister.Track,
            2 => FdcRegister.Sector,
            3 => FdcRegister.Data,
            _ => throw new ArgumentOutOfRangeException(nameof(address), address, "Unsupported FDC register address.")
        };
    }

    private sealed record ControllerBinding(IFdcController Controller, ITimedFdcController TimedController);
}
